File: c/src/dp/creational/singleton.c
#include "singleton.h"

Singleton* instance = NULL;

static Singleton* getInstance() {
  if (instance == NULL) {
    instance = (Singleton*) malloc(sizeof(Singleton));
    instance->valore = 0;
  }
  return instance;
}

static void destroyInstance() {
  if (instance != NULL) {
    free(instance);
    instance = NULL; // dangling pointer
  }
}

void test_singleton() {
  Singleton* s1 = getInstance();
  s1->valore = 42;
  Singleton* s2 = getInstance();
  printf("Valore di s2: %d\n", s2->valore);
  if (s1 == s2) { printf("s1 e s2 puntano alla stessa istanza\n");  }
  destroyInstance();
}

File: rust/src/dp/creational/singleton.rs
use std::sync::{Mutex, OnceLock};

struct Singleton {
  valore: Mutex<i32>,
}

static INSTANCE: OnceLock<Singleton> = OnceLock::new();

impl Singleton {
  fn get_instance() -> &'static Singleton {
    INSTANCE.get_or_init(|| Singleton {
      valore: Mutex::new(0),
    })
  }

  fn set_valore(new_val: i32) {
    let instance = Self::get_instance();
    let mut valore = instance.valore.lock().unwrap();
    *valore = new_val;
  }

  fn get_valore() -> i32 {
    let instance = Self::get_instance();
    *instance.valore.lock().unwrap()
  }
}

pub fn test_singleton() {
  let s1 = Singleton::get_instance();
  Singleton::set_valore(42);

  let s2 = Singleton::get_instance();
  println!("Valore di s2: {}", Singleton::get_valore()); // Deve stampare 42

  if std::ptr::eq(s1, s2) {
    println!("s1 e s2 puntano alla stessa istanza");
  }
}
